### PR TITLE
Add Toolbar Support

### PR DESCRIPTION
**Pull request by Anatidae-Project**
This PR adds the new Toolbar widget support to Ti.DrawerLayout.

The manifest must have a NoActionBar theme for this to work. Otherwise Titanium use use the old ActionBar and prevent this Toolbar from activating. 

```
<application android:theme="@style/Theme.AppCompat.Light.NoActionBar" />
```

You can also hide the toolbar at runtime (needs to be fully tested).

One set, this should work just as a normal Actionbar but with cool Material styling designs!!

### CODE REVIEW

**Comment by manumaticx** [comment]
First of all, thank you so much. You have done an incredible job on this and I really appreciate the good work. :+1: 

I'm going to test this tomorrow. But looking at the code changes, here are some thoughts:

Would it be possible to create the Toolbar at runtime instead of inflating from layout? That way you could only create it if needed and not only setting the visibility to gone. I think turning the toolbar on and off is not necessary after creation. So maybe the best solution (if possible) would be determining wether we need a toolbar and add it to the layout dynamically. Can we determine this by checking the activity's Theme? Otherwise this could also be a creation-property.

Another question I have is what are the downsides of replacing TiCompositeLayout in favor of RelativeLayout? I know, the Toolbar is not working in TiCompositeLayout (I've tested this before). This is something I'll test tomorrow since I'm afraid MapView is not working within RelativeLayout.

Before merging, can you please revert the last commit? - I don't want to change the existing 1.3.5 package. We'll update this to 1.4.0 once everything is tested and working :)

**Comment by manumaticx** [comment]
@Anatidae-Project I have another question. In order to comply with the material design guide for the [Navigation Drawer](https://www.google.com/design/spec/patterns/navigation-drawer.html#) there are some attributes I need to set on the Toolbar like `android:theme` or `android:paddingTop`. This is probably not possible dynamically. So my question is, do you think there is a way to let the user define the Toolbar in a custom xml layout that is used in this module?

**Comment by Anatidae-Project** [comment]
I think you are right about the map. I don't think the map can be in anything but a frame layout. We just did a native project with the map and I'll ask my colleague what she had to do. 

So, the thing about titanium is that I'm guessing it is traversing all the children in the view and applying layout properties to it. The problem is that the base view type for that Titanium container is different than the base view for the toolbar, so that crashes. 

We might have to wrap the toolbar in some sort of titanium proxy that allows for those sizing properties but still provides access to the toolbar. 

I've never added a toolbar programmatically, but there is no reason at all you can't make a separate XML file and inflate that or create an instance completely in code and drop it in. If I remember correctly, the toolbar is based on a linearlayout (but not positive). You can put your own child views in the toolbar.

Here is an example:
```
    <android.support.v7.widget.Toolbar
        android:id="@+id/toolbar"
        android:layout_width="match_parent"
        android:layout_height="?attr/actionBarSize"
        android:background="@color/md_white_1000"
        android:elevation="4dp"
        android:theme="@style/ThemeOverlay.AppCompat.ActionBar"
        app:popupTheme="@style/ThemeOverlay.AppCompat.Light">

        <RelativeLayout
            android:id="@+id/toolbar_logo"
            android:layout_width="match_parent"
            android:layout_height="match_parent"
            android:layout_gravity="center"
            android:paddingBottom="19dp"
            android:paddingLeft="0dp"
            android:paddingRight="30dp"
            android:paddingTop="19dp"
            android:visibility="gone">

            <ImageView
                android:layout_width="match_parent"
                android:layout_height="match_parent"
                android:layout_gravity="center"
                android:scaleType="fitCenter"
                android:src="@drawable/logo" />
        </RelativeLayout>

        <Spinner
            android:id="@+id/spinner_nav"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:visibility="visible" />
```

This was something where I would turn on the logo, making the title "" (so you don't see that). I also have a spinner I turn on and off for the navigation. This is so I can have a pull down list instead of the normal title. 

In the end, they are all views. So it might be to get it perfect would just mean creating a ToolbarProxy for Titanium so that when Titanium aggressively applies layout properties they don't fail. As long as you have the toolbar methods in there, you can add views to the toolbar itself. Or even just have the toolbar the bottom layer in that proxy container and put views on top of the toolbar.

**Comment by Anatidae-Project** [comment]
Oh, here is something too. 

I've used maps plenty of times in fragments adding to a drawer layout. The layout was a relative view. The fragment itself had a framelayout to hold the map (I think). But that works fine.

I don't have much time for testing with deadlines and all. So I can't get a map up soon and run it through it's paces. But we are testing this patch in a beta app today extensively. No map, but we do use webviews. Those have had similar problems to the map module in some regards.

**Comment by manumaticx** [comment]
The more I test, try and think about it the more concerns I have about integrating Toolbar support into the module. Even if it works fine the way you've implemented it but I'm not sure anymore if this should really be part of the Drawer module. The aim of Ti.DrawerLayout is providing a Proxy for the DrawerLayout and it actually shouldn't do much more. Nevertheless, it should be able to support the Toolbar. But it should do this by integrating it in someway. I don't like the approach of creating an instance in the layout of the drawer. This limits the user somehow. 

I also think the toolbar support for Titanium should be accomplished by either a proxy in the SDK or a separate module. This is what I proposed in the according [jira ticket](https://jira.appcelerator.org/browse/TIMOB-17964).

I've also just tested the mapView with your pull request and it works! I actually don't know why, but it's working fine :)

I also successfully separated the toolbar definition in the xml layout but could not yet find a way to define it outside of the module (in the app res). But I'll continue testing this. 

If we can achieve hooking in the Toolbar without touching the drawer module, this would be awesome! I'm also going to try to support the Toolbar in TiComositeLayout.

**Comment by manumaticx** [comment]
@Anatidae-Project Hey, I've just tried putting the toolbar straight on the TiCompositeLayout and that worked !! :open_mouth:

Now, I'm going to try to swap this out into a separate proxy.

**Comment by Anatidae-Project** [comment]
Honestly, the toolbar has to be part of the module layout. It's the only way the drawer can be over the toolbar. 

If Titanium sets a toolbar in the main activity, they would have to put it in some layout there. Upon which you couldn't easily replace it anymore. 

The best thing the Titanium could do would be to wrap a proxy of the toolbar and let you attach it to your main layout then set it as the toolbar. The thing is, once you set an actionbar/toolbar for an activity, it's set. My first attempt was to try unsetting or resetting the the Actionbar in Titanium. I also searched for a way to unset an actionbar. Basically, it just isn't done. There is stuff about hiding it, but that's just visibility.

So basically, right now it's best to think of the toolbar as something you set once per activity and then you are stuck with it. 

Keep in mind that when you set the toolbar is fills the reference when you do Activity.getSupportActionbar(). So you can set it anywhere and anytime, and from that point on supportActionBar is now that. Ideally, Titanium will never put in a default toolbar but rather just let you add one to your layout and set it as the main actionbar. 

Thus, it ends up being in your module still :)

**Comment by manumaticx** [comment]
I totally agree the toolbar has to be part of the drawer layout but I thought it should be added to it rather than defined within it. 

Let's assume Titanium supports the Toolbar and lets you add it to your layout like this:
```js
var win = Ti.UI.createWindow();
var toolbar = Ti.UI.Android.createToolbar();
win.add(toolbar);
win.open();
```

Why shouldn't it be possible to add it to the drawer instead?
```js
var win = Ti.UI.createWindow();
var TiDrawerLayout = require('com.tripvi.drawerlayout');
var toolbar = Ti.UI.Android.createToolbar();
var drawer = TiDrawerLayout.createDrawer();
drawer.addToolbar(toolbar);
win.add(drawer);
win.open();
```

**Comment by Anatidae-Project** [comment]
Thinking further, I don't see a good way around the general toolbar layout. You need your main containing view, then you want to have the toolbar and below it your container you'll be attaching views or fragments to. Which is what my PR had.

Here are some options:
1. Have two containers. One for the toolbar and one for the main view. You have to make a toolbar proxy class that extends toolbar and probably implements whatever TiComositeLayout is using for it's layout.parameter type. That is if you want to let Titanium add viewProxys to the toolbar container as well. Then you can supply a default/standard toolbar instance or allow the user to supply their own toolbar instance provided from another module. 

2. Just keep the layout similar to what is up there. The only difference is that you can let the user supply a different toolbar reference upon which you just replace it with what you have in the layout. This is easier, but doesn't allow for any Titanium viewProxies to live in the toolbar space.

Keep in mind that setting a view to GONE like this isn't going to cost any performance. The size footprint it so small in this particular case too.

The thing I keep coming back to is that Titanium makes it so limiting to do much with the toolbar, this solution is pretty good for most Ti users. If anyone wants to do anything special with the toolbar, then they need to make a module. At that point I'd rather see people branch this module. 

My point is, until Ti 5.x.x and maybe beyond, this is probably the best solution anyway and will service the vast majority of users. It's not like anyone losses anything they already had. 

As a sidenote, my native projects use this guy's drawer:
https://github.com/mikepenz/MaterialDrawer
This is a great open source project. This is a drawer that should be wrapped into this module. We need Titanium to up the support lib or pull from an older version of the drawer as Mike keeps it current. But it offers the fix to the keyboard not pushing up views properly and dismissing when drawer opens. It has a header and footer space you could drop a TiComositeLayout in. The main area can take a custom view too, but as an option the methods he has could be wrapped.

It's a project with someone who doesn't have looming deadlines like myself :)

**Comment by Anatidae-Project** [comment]
Oh, in your code example.... there will need to be something, somewhere, that is like:

```
Ti.UI.Android.setSupportActionBar(toolbar)
```
That would then call Activity.setSupportActionBar() somewhere.

**Comment by manumaticx** [comment]
I thought `Activity.setSupportActionBar()` can be called in the constructor of the Toolbar proxy. I'm currently testing this while creating such a proxy

**Comment by Anatidae-Project** [comment]
I crashed the app when I tried to set a toolbar before adding to a view in another project. But this was right when the toolbar came out... So, maybe it will work.

**Comment by manumaticx** [comment]
The good thing is that users keep control about the toolbar appearance (color, title, menu, actions) ...this is absolutely compatible. So maybe merging this PR won't break everything. But I want to put to the acid test first.

**Comment by manumaticx** [comment]
And again, thanks a lot for all your work and input here! Thta's awesome :)
